core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -141,41 +141,43 @@ private void removeRecovery(ShardRouting routing) {
         addRecovery(routing, false, null);
     }
 
-    private void addInitialRecovery(ShardRouting routing, RoutingTable routingTable) {
-        addRecovery(routing, true, routingTable);
+    private void addInitialRecovery(ShardRouting routing, ShardRouting initialPrimaryShard) {
+        addRecovery(routing, true, initialPrimaryShard);
     }
 
-    private void addRecovery(final ShardRouting routing, final boolean increment, final RoutingTable routingTable) {
+    private void addRecovery(final ShardRouting routing, final boolean increment, final ShardRouting initialPrimaryShard) {
         final int howMany = increment ? 1 : -1;
         assert routing.initializing() : "routing must be initializing: " + routing;
+        assert initialPrimaryShard == null || initialPrimaryShard.assignedToNode() :
+            "shard is initializing but its primary is not assigned to a node";
+
         Recoveries.getOrAdd(recoveriesPerNode, routing.currentNodeId()).addIncoming(howMany);
-        final String sourceNodeId;
-        if (routing.relocatingNodeId() != null) { // this is a relocation-target
-            sourceNodeId = routing.relocatingNodeId();
-            if (routing.primary() && increment == false) { // primary is done relocating
+
+        if (routing.isPeerRecovery()) {
+            // add/remove corresponding outgoing recovery on node with primary shard
+            final ShardRouting primary;
+            if (initialPrimaryShard != null) {
```

Comment:
nit: can we move the look up of the primary to the callers that pass null? this method is hairy enough :)